system/libuv/libuv/Makefile

```diff
@@ -0,0 +1,107 @@
+# this work for additional information regarding copyright ownership.  The
+# ASF licenses this file to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance with the
+# License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+# WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
+# License for the specific language governing permissions and limitations
+# under the License.
+#
+############################################################################
+
+include $(APPDIR)/Make.defs
+
+LIBUV_PATCHS ?= $(sort $(wildcard $(LIBUV_DIR)/000*.patch))
+WGET ?= wget
+UNPACK ?= unzip -q
+
+$(LIBUV_DIR)/$(CONFIG_LIBUV_TARBALL):
+	@echo "Downloading: $(CONFIG_LIBUV_TARBALL)"
+	$(Q) $(WGET) $(CONFIG_LIBUV_URL)/$(CONFIG_LIBUV_TARBALL) \
+		-O $(LIBUV_DIR)/$(CONFIG_LIBUV_TARBALL)
+
+$(LIBUV_UNPACKDIR): $(LIBUV_DIR)/$(CONFIG_LIBUV_TARBALL)
+	@echo "Unpacking: $(CONFIG_LIBUV_TARBALL) -> $(LIBUV_UNPACKDIR)"
+	$(call DELDIR, $(LIBUV_UNPACKDIR))
+	$(Q) unzip -q -d "$(LIBUV_DIR)" "$(LIBUV_DIR)/$(CONFIG_LIBUV_TARBALL)"
+	$(Q) cat $(LIBUV_PATCHS) | \
+		patch -s -N -d $(LIBUV_UNPACKDIR) -p1
+	$(Q) touch $(LIBUV_UNPACKDIR)
+
+# Build libuv library
+
+CFLAGS += -I$(LIBUV_UNPACKDIR)/src
+CFLAGS += -I$(LIBUV_UNPACKDIR)/src/unix
+CFLAGS += -D__NUTTX__
+
+CSRCS += $(LIBUV_UNPACKDIR)/src/unix/core.c
+CSRCS += $(LIBUV_UNPACKDIR)/src/unix/poll.c
+CSRCS += $(LIBUV_UNPACKDIR)/src/unix/loop.c
+CSRCS += $(LIBUV_UNPACKDIR)/src/unix/thread.c
+CSRCS += $(LIBUV_UNPACKDIR)/src/unix/no-proctitle.c
+ifeq ($(CONFIG_LIBUV_LOOP_WATCHERS),y)
+CSRCS += $(LIBUV_UNPACKDIR)/src/unix/loop-watcher.c
+endif
+
+# FIXME signal does not work yet
+ifeq ($(CONFIG_LIBUV_SIGNAL),y)
+CSRCS += $(LIBUV_UNPACKDIR)/src/unix/signal.c
+endif
+
+# FIXME process does not work yet
+ifeq ($(CONFIG_LIBUV_PROCESS),y)
+CSRCS += $(LIBUV_UNPACKDIR)/src/unix/process.c
+endif
+
+CSRCS += $(LIBUV_UNPACKDIR)/src/unix/nuttx.c
+
+ifeq ($(CONFIG_LIBUV_STREAM),y)
+CSRCS += $(LIBUV_UNPACKDIR)/src/unix/nuttx_stream.c
+endif
+
+ifeq ($(CONFIG_LIBUV_TCP),y)
+CSRCS += $(LIBUV_UNPACKDIR)/src/unix/nuttx_tcp.c
+endif
+
+ifeq ($(CONFIG_LIBUV_WQ),y)
+CSRCS += $(LIBUV_UNPACKDIR)/src/unix/nuttx_threadpool.c
+endif
+
+ifeq ($(CONFIG_LIBUV_ASYNC),y)
+CSRCS += $(LIBUV_UNPACKDIR)/src/unix/async.c
+endif
+
+ifeq ($(CONFIG_LIBUV_TIMER),y)
+ifeq ($(CONFIG_LIBUV_TIMER_NUTTX),y)
+CSRCS += $(LIBUV_UNPACKDIR)/src/unix/nuttx_timer.c
+else
+CSRCS += $(LIBUV_UNPACKDIR)/src/timer.c
+endif
+endif
+
+CSRCS += $(LIBUV_UNPACKDIR)/src/uv-common.c
+CSRCS += $(LIBUV_UNPACKDIR)/src/strscpy.c
+
+ifeq ($(CONFIG_LIBUV_NET),y)
+CSRCS += $(LIBUV_UNPACKDIR)/src/inet.c
+endif
+
+context:: $(LIBUV_UNPACKDIR)
+
+clean::
+	$(call DELFILE, $(OBJS))
```

Comment:
Is this necessary?  Application.mk already deletes object files.  If the object files are generated else where then you need to extend clean::.

Reply:
my understanding is CLEAN only removes "$(Q) rm -f *$(OBJEXT) *$(LIBEXT)", so "*.o" and "*.a".
It can be tricky to use VPATH here due to potential file name collisions between libuv/{libuv,tests} and libuv/libuv-1.38.1/.
So I added the files in CSRCS variable with absolute path. Hence the CLEAN target does not work anymore because the ".o" files are located in another folder, "libuv-1.38.1/src/[unix]".

If it's an issue, I can switch to VPATH and ensure there is no filename collision.

Reply:
It's not an issue it's exactly how it should work.  But are your $(LIBUV_UNPACKDIR)/src/*.o files being cleaned? As I see it, it cleans only at the libuv/libuv level.

Reply:
Yes, all the objects are cleaned with this trick, even the ones not in libuv/ or tests/ folders like "./libuv-1.38.1/src/unix/nuttx_stream.path.to.nuttx.incubator-nuttx-apps.system.libuv.libuv.o".
OBJS is defined based on $(CSRCS:.c=$(SUFFIX)$(OBJEXT)) so it works fine.
How do you feel about this ?
What would be the correct solution to ensure all the objects are cleaned without extending the clean:: target ?

Reply:
No. You are right.
The clean target called from Application only removes the files at the current level, it doesn't clean subdirectories, we need to extend it to clean any other file generated. The way you extended it to clean all OBJS does go through all the subdirectories.

Reply:
The CLEAN operation does more than just remove object files.  See definitions in tools/Config.mk

Reply:
The CLEAN operation of tools/Config.mk is also over-ridden by architecture-specific operations that generate other kinds of files.  See for example, tools/zds/Config.mk.  If you simply remove the $(OBJS) in the sub-directory, the clean will be incomplete in some cases.

Reply:
The clean target as called from the general Application.mk does the regular CLEAN.  The application specific Makefile extends that to clean any other files that can not be reached by CLEAN.